pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -262,40 +252,58 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListeners(env *model.En
 			continue
 		}
 		listenerType = plugin.ModelProtocolToListenerProtocol(protocol)
+		allChains := []plugin.FilterChain{}
+		var httpOpts *httpListenerOpts
+		var tcpNetworkFilters []listener.Filter
 		switch listenerType {
 		case plugin.ListenerProtocolHTTP:
-			httpOpts := &httpListenerOpts{
+			httpOpts = &httpListenerOpts{
 				routeConfig:      configgen.buildSidecarInboundHTTPRouteConfig(env, node, push, instance),
 				rds:              "", // no RDS for inbound traffic
 				useRemoteAddress: false,
 				direction:        http_conn.INGRESS,
 			}
-			if listenerOpts.tlsMultiplexed {
-				listenerOpts.filterChainOpts = []*filterChainOpts{
-					{
-						httpOpts:          httpOpts,
-						transportProtocol: authn.EnvoyRawBufferMatch,
-					}, {
-						httpOpts:          httpOpts,
-						transportProtocol: authn.EnvoyTLSMatch,
-					},
-				}
-			} else {
-				listenerOpts.filterChainOpts = []*filterChainOpts{
-					{
-						httpOpts: httpOpts,
-					},
-				}
-			}
 		case plugin.ListenerProtocolTCP:
-			listenerOpts.filterChainOpts = []*filterChainOpts{{
-				networkFilters: buildInboundNetworkFilters(instance),
-			}}
+			tcpNetworkFilters = buildInboundNetworkFilters(instance)
 
 		default:
 			log.Warnf("Unsupported inbound protocol %v for port %#v", protocol, instance.Endpoint.ServicePort)
 			continue
 		}
+		for _, p := range configgen.Plugins {
+			params := &plugin.InputParams{
+				ListenerProtocol: listenerType,
+				Env:              env,
+				Node:             node,
+				ProxyInstances:   proxyInstances,
+				ServiceInstance:  instance,
+				Port:             endpoint.ServicePort,
+			}
+			chains := p.OnInboundFilterChains(params)
+			if len(chains) == 0 {
+				continue
+			}
+			if len(allChains) != 0 {
+				log.Errorf("Found two plugin returns non empty filter chains!")
+				allChains = []plugin.FilterChain{}
+				break
+			}
+			allChains = chains
```

Comment:
I dont understand this logic: It seems you want to make sure only one plugin returns filter chains
and then outside for loop, you are iterating over all the chains which implies you are prepared for multiple filter chains.. So why not remove that check of len(allchains) != 0.. and simply do 
`allChains = append(allChains, chains…)`

Reply:
Yes. Just remove this check and append. If plugins mess up, then we have a bigger problem.